{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import re\n",
    "from pathlib import Path\n",
    "\n",
    "from IPython.display import display, HTML, Markdown\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scipy.sparse as sp_sparse\n",
    "\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import tables\n",
    "\n",
    "# Project level imports\n",
    "from larval_gonad.notebook import Nb\n",
    "from larval_gonad.io import cellranger_umi\n",
    "from larval_gonad.config import memory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "NUCS = ['A', 'C', 'G', 'T']\n",
    "\n",
    "\n",
    "def decompress_seq(x: int, length=16):\n",
    "    \"\"\" Un-pack a DNA sequence from a 2-bit format\n",
    "\n",
    "    Based on code from: https://github.com/10XGenomics/cellranger\n",
    "\n",
    "    cellranger/lib/python/cellranger/utils.py\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    x : int\n",
    "        Number sequence to be decoded.\n",
    "    length : int\n",
    "        Length of the barcode. This can be found in the molecular info hdf5\n",
    "        file from 10x genome.\n",
    "        molInfo.get_node_attr('/metrics', 'chemistry_barcode_read_length')\n",
    "\n",
    "    \"\"\"\n",
    "    bits = 64\n",
    "    x = np.uint64(x)\n",
    "    assert length <= (bits/2 - 1)\n",
    "    if x & (1 << (bits-1)):\n",
    "        return 'N' * length\n",
    "    result = bytearray(length)\n",
    "    for i in range(length):\n",
    "        result[(length-1)-i] = bytearray(NUCS[x & np.uint64(0b11)].encode())[0]\n",
    "        x = x >> np.uint64(2)\n",
    "    return result.decode()\n",
    "\n",
    "\n",
    "def two_bit_mapper(iterable):\n",
    "    \"\"\"Return a dictionary mapping 2bit encoded Seqs.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    iterable : list-like\n",
    "        Unique list of 2bit encoded sequences.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    dict : Mapper from encoded to decoded\n",
    "\n",
    "    \"\"\"\n",
    "    return {k: decompress_seq(k) for k in iterable}\n",
    "\n",
    "\n",
    "def decode_cell_names(iterable):\n",
    "    \"\"\"Use two_bit_mapper to decode cell names.\n",
    "\n",
    "    iterable : np.array\n",
    "        An array of twobit encoded cell names.\n",
    "\n",
    "    \"\"\"\n",
    "    mapper = two_bit_mapper(np.unique(iterable))\n",
    "    return [mapper[x] for x in iterable]\n",
    "\n",
    "\n",
    "def cellranger_umi(fname):\n",
    "    with tables.open_file(fname, 'r') as f:\n",
    "        group = f.get_node('/')\n",
    "        cell_ids = getattr(group, 'barcode').read()\n",
    "        umi = getattr(group, 'umi').read()\n",
    "        gene = getattr(group, 'gene').read()\n",
    "        read_cnts = getattr(group, 'reads').read()\n",
    "        gene_ids = getattr(group, 'gene_ids').read()\n",
    "\n",
    "    # gene_ids: np.array\n",
    "    gene_ids = np.append(gene_ids, ['Not a Gene'])\n",
    "    cell_names = decode_cell_names(cell_ids)\n",
    "    umi = decode_cell_names(umi)\n",
    "    gene = [gene_ids[x] for x in gene]\n",
    "\n",
    "    return pd.DataFrame(dict(\n",
    "        cell_id=cell_names,\n",
    "        umi=umi,\n",
    "        gene=gene,\n",
    "        read_cnt=read_cnts\n",
    "    ))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "rep = 1\n",
    "umi = cellranger_umi(f'../output/scrnaseq-wf/scrnaseq_samples/testis{rep}_force/outs/molecule_info.h5')\n",
    "umi = umi.query('gene != \"Not a Gene\"')\n",
    "umi.to_csv(f'../output/notebook/2018-11-01_make_stack_for_salsa_rep{rep}.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "rep = 2\n",
    "umi = cellranger_umi(f'../output/scrnaseq-wf/scrnaseq_samples/testis{rep}_force/outs/molecule_info.h5')\n",
    "umi = umi.query('gene != \"Not a Gene\"')\n",
    "umi.to_csv(f'../output/notebook/2018-11-01_make_stack_for_salsa_rep{rep}.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "rep = 3\n",
    "umi = cellranger_umi(f'../output/scrnaseq-wf/scrnaseq_samples/testis{rep}_force/outs/molecule_info.h5')\n",
    "umi = umi.query('gene != \"Not a Gene\"')\n",
    "umi.to_csv(f'../output/notebook/2018-11-01_make_stack_for_salsa_rep{rep}.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:larval_gonad]",
   "language": "python",
   "name": "conda-env-larval_gonad-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
