---
title: Combining scRNA-Seq
author: Justin Fear
date: 'Compiled: `r format(Sys.Date(), "%B %d, %Y")`'
---

```{r, include=FALSE}
knitr::opts_chunk$set(collapse=TRUE, warning=FALSE, message=FALSE,
                      bootstrap.show.code=FALSE, bootstrap.show.output=FALSE,
                      fig.ext='png')
```

```{r setup}
options(repr.plot.width=10, repr.plot.height=10)
source('../lib/seurat.R')
library(Seurat)
library(dplyr)
library(Matrix)

OUTDIR <- '../output/scrnaseq_combine_force'
dir.create(file.path(OUTDIR), showWarnings = FALSE)
REFERENCES_DIR <- Sys.getenv('REFERENCES_DIR')
params <- c(0.4, 0.6, 1.0)

# Get list of mitochondiral genes
fbgn2chrom <- read.table('../output/fbgn2chrom.tsv', header=T)
fbgn2symbol <- read.csv(file.path(REFERENCES_DIR, 'dmel/r6-16/fb_annotation/dmel_r6-16.fb_annotation'), header=T, sep = '\t')[, c('gene_symbol', 'primary_FBgn')]
mito <- fbgn2chrom[fbgn2chrom$chrom == 'chrM', 'FBgn']
```

```{r load.data, cache=TRUE}
print('Loading Data')
load.data <- function(fname, name){
	# Load the 10x dataset
	tenX.data <- Read10X(data.dir = fname)
	obj <- CreateSeuratObject(raw.data = tenX.data, min.cells = 3, min.genes = 200, project = 'GroupCluster')
	obj@meta.data$rep <- name
	obj <- FilterCells(object = obj, subset.names = c("nGene"), low.thresholds = c(200), high.thresholds = c(6000))
	obj <- NormalizeData(object = obj, normalization.method = "LogNormalize", scale.factor = 1e4)
	obj <- ScaleData(object = obj, vars.to.regress = c("nUMI"), display.progress = FALSE)

	# Find the highly variable genes
	obj <- FindVariableGenes(
	    object = obj,
	    mean.function = ExpMean,
	    dispersion.function = LogVMR,
	    do.plot = FALSE,
	    x.low.cutoff = 0.01,
	    x.high.cutoff = 3,
	    y.cutoff = 0.5,
	    y.high.cutoff = Inf
	)

	# Rename cells incase they overlap between replicates
	obj@cell.names <- paste(name, obj@cell.names, sep='_')
	colnames(obj@raw.data) <- paste(name, colnames(obj@raw.data), sep = "_")
	rownames(obj@meta.data) <- paste(name, rownames(obj@meta.data), sep = "_")
	return(obj)
}

r1 <- load.data('../output/testis1_force/outs/filtered_gene_bc_matrices/dm6.16', 'rep1')
r2 <- load.data('../output/testis2_force/outs/filtered_gene_bc_matrices/dm6.16', 'rep2')
r3 <- load.data('../output/testis3_force/outs/filtered_gene_bc_matrices/dm6.16', 'rep3')

```

```{r plot.ngene}
VlnPlot(object = r1, features.plot = c("nGene", "nUMI"), nCol = 2)
VlnPlot(object = r2, features.plot = c("nGene", "nUMI"), nCol = 2)
VlnPlot(object = r3, features.plot = c("nGene", "nUMI"), nCol = 2)
```

## Combined Data

```{r combine, cache=TRUE}
print('Combining Data')
# Create list of genes to use
g.1 <- r1@var.genes
print(paste0('variable genes rep 1: ', length(g.1)))

g.2 <- r2@var.genes
print(paste0('variable genes rep 2: ', length(g.2)))

g.3 <- r3@var.genes
print(paste0('variable genes rep 3: ', length(g.3)))

genes.use <- intersect(intersect(g.1, g.2), g.3)
print(paste0('variable genes intersection: ', length(genes.use)))

combined <- RunMultiCCA(
			list(r1, r2, r3),
			genes.use = genes.use,
			num.cc = 10
			)
```

```{r}
p1 <- DimPlot(object=combined, reduction.use='cca', group.by='rep', pt.size=0.5, do.return=TRUE)
p2 <- VlnPlot(object=combined, features.plot='CC1', group.by='rep', do.return=TRUE)
plot_grid(p1, p2)
```

## Aligned Data

```{r align, cache=TRUE}
print('Aligning Data')
combined <- AlignSubspace(
			  combined,
			  reduction.type = "cca",
			  grouping.var = "rep",
			  dims.align = 1:10
			  )
```

```{r}
p1 <- VlnPlot(object=combined, features.plot='ACC1', group.by='rep', do.return=TRUE)
p2 <- VlnPlot(object=combined, features.plot='ACC2', group.by='rep', do.return=TRUE)
plot_grid(p1, p2)
```

## tSNE at Different Resolutions

```{r, cache=TRUE}
print('Building Clusters')
combined <- RunTSNE(combined, reduction.use = "cca.aligned", dims.use = 1:10, do.fast = T)
combined <- FindClusters(combined, reduction.type = "cca.aligned", resolution = params, dims.use = 1:10)
```

```{r}
combined <- SetAllIdent(combined, id  = 'res.0.4')
p1 <- TSNEPlot(combined, do.return = T, pt.size = 0.5, group.by = "rep")
p2 <- TSNEPlot(combined, do.label = T, do.return = T, pt.size = 0.5)
p <- plot_grid(p1, p2)
title <- ggdraw() + draw_label("res 0.4", fontface='bold')
plot_grid(title, p, ncol=1, rel_heights=c(0.1, 1)) # rel_heights values control title margins
```

```{r}
combined <- SetAllIdent(combined, id  = 'res.0.6')
p1 <- TSNEPlot(combined, do.return = T, pt.size = 0.5, group.by = "rep")
p2 <- TSNEPlot(combined, do.label = T, do.return = T, pt.size = 0.5)
p <- plot_grid(p1, p2)
title <- ggdraw() + draw_label("res 0.6", fontface='bold')
plot_grid(title, p, ncol=1, rel_heights=c(0.1, 1)) # rel_heights values control title margins
```

```{r}
combined <- SetAllIdent(combined, id  = 'res.1.0')
p1 <- TSNEPlot(combined, do.return = T, pt.size = 0.5, group.by = "rep")
p2 <- TSNEPlot(combined, do.label = T, do.return = T, pt.size = 0.5)
p <- plot_grid(p1, p2)
title <- ggdraw() + draw_label("res 1.0", fontface='bold')
plot_grid(title, p, ncol=1, rel_heights=c(0.1, 1)) # rel_heights values control title margins
```

## Save Results

```{r}
print('Dumping Seurat Object')
dump_seurat(object = combined, dir = OUTDIR)
```

```{r}
print('Saving Biomarkers')
# Save cluster info
for (i in params) {
    name <- paste0('res.', i)
    fname <- paste0('biomarkers_', i, '.tsv')
    combined <- SetAllIdent(combined, id = name)
    markers <- FindAllMarkers(object = combined, only.pos = TRUE, min.pct = 0.25, thresh.use = 0.25, print.bar = FALSE)
    markers = merge(fbgn2symbol, markers, by.x='primary_FBgn', by.y='gene', all.y=T)
    save_biomarkers(markers = markers, dir = OUTDIR, fname = fname)
}
```
