---
title: Germcells vs Soma
author: Justin Fear
---

I would like to do a differential expression analysis of germcell vs somatic
cells. This is pretty staright forward given a seurot object. I just need to
load the object and then run the findmarkers with the correct groups.

I am assuming the following annotations:

* 0: 'Late 1ยบ Spermatocytes (0)',
* 1: 'Late Cyst Cells (1)',
* 2: 'Spermatogonia (2)',
* 3: 'Early 1ยบ Spermatocytes (3)',
* 4: 'Late Cyst Cells (4)',
* 5: 'Late Cyst Cells (5)',
* 6: 'Early Cyst Cells (6)',
* 7: 'Terminal Epithelium(7)',
* 8: 'Pigment (8)',
* 9: 'Unknown (9)',


```{r load_libs}
library(Seurat)
library(dplyr)
source('../lib/seurat.R')

REFERENCES_DIR <- Sys.getenv('REFERENCES_DIR')
OUTDIR <- '../output/scrnaseq_combine_force'
fbgn2symbol <- read.csv(file.path(REFERENCES_DIR, 'dmel/r6-16/fb_annotation/dmel_r6-16.fb_annotation'), header=T, sep = '\t')[, c('gene_symbol', 'primary_FBgn')]
```

```{r load.data}
load('../output/scrnaseq_combine_force/seurat.Robj')
sobj <- object
```

```{r load.resolution}
sobj <- SetAllIdent(sobj, id='res.0.4')
```

```{r find.markers}
markers <- FindMarkers(sobj, ident.1 = c(0, 2, 3), ident.2 = c(1, 4, 5, 6))
```

```{r write.output}
fname <- '2018-05-10_scrnaseq_germ_vs_soma_biomarkers.tsv'
markers$gene = rownames(markers)
markers = merge(fbgn2symbol, markers, by.x='primary_FBgn', by.y='gene', all.y=T)
save_biomarkers(markers = markers, dir = OUTDIR, fname = fname)
```

