"""Scrip to generate figures and needed material for the Science Submssion"""
import os
import yaml
from pathlib import Path

import pandas as pd

from larval_gonad.config import read_config

workdir: "."
configfile: "../config/common.yaml"
colors_config = read_config("../config/colors.yaml")

SAMPLES = ["testis1", "testis2", "testis3"]

REFERENCES_DIR = os.environ.get("REFERENCES_DIR", config.get("REFERENCES_DIR", ""))
ASSEMBLY = config["assembly"]
TAG = config["tag"]

LIT_GENES = read_config("../config/literature_genes.yaml")
PTRAP_GENES = read_config("../config/ptrap_genes.yaml")
COLORS = yaml.full_load(open("../config/colors.yaml"))

wildcard_constraints: 
    sample="testis\d"

rule all:
    input: 
        # expand("../output/science_submission/figures/lit_genes_umap_panel_{group}.svg", group=LIT_GENES.keys()),
        # expand("../output/science_submission/figures/ptrap_genes_umap_panel_{group}.svg", group=PTRAP_GENES.keys()),
        # expand("../output/science_submission/figures/lit_genes_expression_patterns_panel_{group}.svg", group=LIT_GENES.keys()),
        # expand("../output/science_submission/figures/ptrap_genes_expression_patterns_panel_{group}.svg", group=PTRAP_GENES.keys()),
        "../output/science_submission/figures/cluster-legend.svg",
        "../output/science_submission/figures/lit-gene-evidence.svg",
        "../output/science_submission/figures/lit-gene-evidence_soma.svg",
        expand(
            "../output/science_submission/figures/GvLPS_{direction}_biased_chrom_location.svg",
            direction=['G', 'LPS']
        ),
        expand(
            "../output/science_submission/figures/{clus}_chrom_enrichment.svg",
            clus=["G", "EPS", "MPS", "LPS"]
        ),
        expand(
            "../output/science_submission/figures/GvLPS_{direction}_expression_by_chrom.svg",
            direction=['G', 'LPS']
        ),
        "../output/science_submission/figures/msci_panel.svg",
        "../output/science_submission/biomarkers.tsv",
        "../output/science_submission/figures/ptrap_expression_panel.svg",
        "../output/science_submission/figures/chrom_counts_panel.svg"


################################################################################
# Part 1: scRNA-Seq intron, cluster, annotate, and validate
################################################################################
rule plot_cluster_legend:
    params:
        cluster_colors=COLORS["clusters"],
        legend_names=config["legend_names"]
    output: "../output/science_submission/figures/cluster-legend.svg"
    script: "scripts/plot_cluster_legend.py"


rule lit_gene_umap_panel:
    input: f"../references/gene_annotation_dmel_{TAG}.feather"
    params: 
        pattern=lambda wildcards: "../output/seurat3-cluster-wf/combined_n3_figures/gene_projections/{symbol}_{FBgn}.svg",
        genes=lambda wildcards: LIT_GENES[wildcards.group]
    output: "../output/science_submission/figures/lit_genes_umap_panel_{group}.svg"
    script: "scripts/gene_panels.py"


rule lit_gene_expression_patterns_panel:
    input: f"../references/gene_annotation_dmel_{TAG}.feather"
    params: 
        pattern=lambda wildcards: "../output/seurat3-cluster-wf/combined_n3_figures/expression_patterns/{symbol}_{FBgn}.svg",
        genes=lambda wildcards: LIT_GENES[wildcards.group]
    output: "../output/science_submission/figures/lit_genes_expression_patterns_panel_{group}.svg"
    script: "scripts/gene_panels.py"


rule ptrap_gene_umap_panel:
    input: f"../references/gene_annotation_dmel_{TAG}.feather"
    params: 
        pattern=lambda wildcards: "../output/seurat3-cluster-wf/combined_n3_figures/gene_projections/{symbol}_{FBgn}.svg",
        genes=lambda wildcards: PTRAP_GENES[wildcards.group]
    output: "../output/science_submission/figures/ptrap_genes_umap_panel_{group}.svg"
    script: "scripts/gene_panels.py"


rule ptrap_gene_expression_patterns_panel:
    input: f"../references/gene_annotation_dmel_{TAG}.feather"
    params: 
        pattern=lambda wildcards: "../output/seurat3-cluster-wf/combined_n3_figures/expression_patterns/{symbol}_{FBgn}.svg",
        genes=lambda wildcards: PTRAP_GENES[wildcards.group]
    output: "../output/science_submission/figures/ptrap_genes_expression_patterns_panel_{group}.svg"
    script: "scripts/gene_panels.py"


rule plot_lit_gene_evidence:
    input:
         gene_metadata = f"../references/gene_annotation_{ASSEMBLY}_{TAG}.feather",
         lit_evidence = '../data/external/miriam/lit_gene_dummy_vars.tsv',
         zscore_by_cluster_rep="../output/seurat3-cluster-wf/zscore_by_cluster_rep.feather"
    output: "../output/science_submission/figures/lit-gene-evidence.svg"
    params: cmap=COLORS["heatmap"]
    script: "scripts/plot_lit_gene_evidence.py"


rule plot_lit_gene_evidence_soma:
    input:
         gene_metadata = f"../references/gene_annotation_{ASSEMBLY}_{TAG}.feather",
         lit_evidence = '../data/external/miriam/lit_gene_dummy_vars.tsv',
         zscore_by_cluster_rep="../output/seurat3-cluster-wf/zscore_by_cluster_rep.feather"
    output: "../output/science_submission/figures/lit-gene-evidence_soma.svg"
    params: cmap=COLORS["heatmap"]
    script: "scripts/plot_lit_gene_evidence_soma.py"


rule plot_deg_chrom_locs:
    input: 
        tpm="../output/seurat3-cluster-wf/tpm_by_cluster_rep.feather",
        deg="../output/seurat3-cluster-wf/germline_deg/GvLPS_{direction}_biased.pkl",
        gene_annot=f"../references/gene_annotation_{ASSEMBLY}_{TAG}.feather"
    output: "../output/science_submission/figures/GvLPS_{direction}_biased_chrom_location.svg"
    script: "scripts/plot_locations.py"


rule plot_chrom_bias:
    input:
        tpm="../output/seurat3-cluster-wf/tpm_by_cluster_rep.feather",
        gene_annot=f"../references/gene_annotation_{ASSEMBLY}_{TAG}.feather"
    output: "../output/science_submission/figures/{clus}_chrom_enrichment.svg"
    script: "scripts/plot_chrom_bias.py"


rule plot_GvLPS_expression_by_chrom:
    input:
        tpm="../output/seurat3-cluster-wf/tpm_by_cluster_rep.feather",
        biased="../output/seurat3-cluster-wf/germline_deg/GvLPS_{direction}_biased.pkl",
        gene_annot=f"../references/gene_annotation_{ASSEMBLY}_{TAG}.feather"
    output: "../output/science_submission/figures/GvLPS_{direction}_expression_by_chrom.svg"
    script: "scripts/plot_GvLPS_expression_by_chrom.py"


rule plot_msci_panel:
    input: 
        autosome_ratios='../output/x-to-a-wf/db/commonly_expressed.dat',
        tpm='../output/seurat3-cluster-wf/tpm_by_cluster.feather',
        commonly_expressed="../output/cellselection-wf/commonly_expressed_genes.pkl",
        gene_annot=f"../references/gene_annotation_{ASSEMBLY}_{TAG}.feather"
    params:
          cluster_color=COLORS["clusters"],
          cluster_order=config['cluster_order'],
    output: "../output/science_submission/figures/msci_panel.svg"
    script: "scripts/msci_panel.py"

rule plot_ptrap_panel:
    input:
        clusters="../output/seurat3-cluster-wf/combined_n3_clusters.feather",
        zscores="../output/seurat3-cluster-wf/zscore_by_cell.feather",
        umap="../output/seurat3-cluster-wf/combined_n3_umap.feather",
        norm="../output/seurat3-cluster-wf/combined_n3_normalized.feather"
    output: "../output/science_submission/figures/ptrap_expression_panel.svg"
    script: "scripts/ptrap_expression_panel.py"

rule plot_chrom_count_panel:
    input: 
        adult_bulk="../output/expression-atlas-wf/w1118_gene_counts.feather",
        L3_bulk="../output/bulk2-rnaseq-wf/testis_ovary_counts.feather",
        L3_sc_agg="../output/seurat3-cluster-wf/aggegated_gene_counts.feather",
        L3_sc_by_class="../output/seurat3-cluster-wf/aggegated_gene_counts_by_germ_soma.feather",
        adult_sc_agg="",
        adult_sc_by_class="",
    output: "../output/science_submission/figures/chrom_counts_panel.svg"
    script: "scripts/plot_chrom_counts_panel.py"

################################################################################
# Tables
################################################################################
rule gene_locs:
    input: Path(REFERENCES_DIR, ASSEMBLY, TAG, "gtf", f"{ASSEMBLY}_{TAG}.gtf.db").as_posix()
    params: chrom_order=config["chrom_order"]
    output: "../output/science_submission/gene_locations.feather"
    script: "scripts/gene_locations.py"


rule raw_to_tsv:
    input: '../output/cellselection-wf/raw.feather'
    output: '../output/science_submission/scRNAseq_raw_counts.tsv.gz'
    run:
        pd.read_feather(input[0]).to_csv(output[0], index=False, sep="\t")


rule seurat_normalized_to_tsv:
    input: '../output/seurat3-cluster-wf/combined_n3_normalized.feather'
    output: '../output/science_submission/scRNAseq_seurat_normalized_counts.tsv.gz'
    run:
        pd.read_feather(input[0]).drop("gene_symbol", axis=1).to_csv(output[0], index=False, sep="\t")


rule raw_by_celltype_to_tsv:
    input: '../output/seurat3-cluster-wf/raw_by_cluster_rep.feather'
    output: '../output/science_submission/scRNAseq_raw_counts_by_celltype_rep.tsv.gz'
    run:
        pd.pivot_table(
            pd.read_feather(input[0]),
            index="FBgn",
            columns=["cluster", "rep"],
            values="UMI",
            aggfunc="first"
        ).to_csv(output[0], sep="\t")


rule tpm_by_celltype_to_tsv:
    input: '../output/seurat3-cluster-wf/tpm_by_cluster_rep.feather'
    output: '../output/science_submission/scRNAseq_tpm_counts_by_celltype_rep.tsv.gz'
    run:
        pd.pivot_table(
            pd.read_feather(input[0]),
            index="FBgn",
            columns=["cluster", "rep"],
            values="TPM",
            aggfunc="first"
        ).to_csv(output[0], sep="\t")


rule zscores_by_celltype_to_tsv:
    input: '../output/seurat3-cluster-wf/zscore_by_cluster_rep.feather'
    output: '../output/science_submission/scRNAseq_zscores_by_celltype_rep.tsv.gz'
    run:
        pd.pivot_table(
            pd.read_feather(input[0]),
            index="FBgn",
            columns=["cluster", "rep"],
            values="zscore",
            aggfunc="first"
        ).to_csv(output[0], sep="\t")

rule biomarkers_to_tsv:
    input: "../output/seurat3-cluster-wf/combined_n3_biomarkers.feather"
    output: "../output/science_submission/biomarkers.tsv"
    params: cluster_annot = config["cluster_annot"]
    run:
        (
            pd.read_feather(input[0])
            .assign(cluster=lambda x: x.cluster.map(lambda y: params.cluster_annot[y]))
            .to_csv(output[0], index=False, sep="\t")
        )
    