---
title: Seurat Combine and Cluster
author: Justin Fear
date: 'Compiled: `r format(Sys.Date(), "%B %d, %Y")`'
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(collapse=TRUE, warning=FALSE, message=FALSE, cache.lazy=FALSE,
                      bootstrap.show.code=FALSE, bootstrap.show.output=FALSE,
                      fig.ext='png')
```

```{r set_globals_snakemake, include=FALSE, eval=FALSE}
get_dir_names <- function(l){
    return(file.path('../', dirname(l)))
}

DATA <- lapply(snakemake@input[["data"]], get_dir_names)
FBGN2CHROM <- file.path('../', snakemake@input[['fbgn2chrom']])
OUTDIR <- file.path('../', dirname(snakemake@output[[1]]))
SAMPLES <- snakemake@params[['samples']]

GENE_LOW <- snakemake@params[['gene_low']]
if (GENE_LOW == 'None'){
    GENE_LOW = -Inf
} else {
    GENE_LOW = as.numeric(GENE_LOW)
}

GENE_HIGH <- snakemake@params[['gene_high']]
if (GENE_HIGH == 'None'){
    GENE_HIGH = Inf
} else {
    GENE_HIGH = as.numeric(GENE_HIGH)
}

UMI_LOW <- snakemake@params[['umi_low']]
if (UMI_LOW == 'None'){
    UMI_LOW = -Inf
} else {
    UMI_LOW = as.numeric(UMI_LOW)
}

UMI_HIGH <- snakemake@params[['umi_high']]
if (UMI_HIGH == 'None'){
    UMI_HIGH = Inf
} else {
    UMI_HIGH = as.numeric(UMI_HIGH)
}
```

```{r set_globals_testing, include=FALSE, eval=TRUE}
DATA <- c(
    '/home/fearjm/local_data_store/larval_gonad/output/cellranger-force-wf/testis1/outs/filtered_gene_bc_matrices/dmelr6-24',
    '/home/fearjm/local_data_store/larval_gonad/output/cellranger-force-wf/testis2/outs/filtered_gene_bc_matrices/dmelr6-24',
    '/home/fearjm/local_data_store/larval_gonad/output/cellranger-force-wf/testis3/outs/filtered_gene_bc_matrices/dmelr6-24'
)
FBGN2CHROM <- '../../output/fbgn2chrom.tsv'
OUTDIR <- '../../output/seurat-cluster-wf/combined'
SAMPLES <- c('testis1', 'testis2', 'testis3')

GENE_LOW <- 200
GENE_HIGH <- Inf

UMI_LOW <- -Inf
UMI_HIGH <- Inf
```

```{r set_other_globas, include=FALSE}
REFERENCES_DIR <- Sys.getenv('REFERENCES_DIR', '/data/LCDB/lcdb-references')
PARAMS <- c(0.4, 0.6, 1)
```

## Combined
Gene Low End Cutoff: `r GENE_LOW`
Gene High End Cutoff: `r GENE_HIGH`
UMI Low End Cutoff: `r UMI_LOW`
UMI High End Cutoff: `r UMI_HIGH`

```{r load_libraries, include=FALSE}
options(repr.plot.width=10, repr.plot.height=10)
source('../../lib/seurat.R')
library(Seurat)
library(dplyr)
library(Matrix)
library(data.table)
library(yaml)
```

```{r import_annotations}
fbgn2chrom <- read.table(FBGN2CHROM, header=T)
fbgn2symbol <- read.csv(file.path(REFERENCES_DIR, 'dmel/r6-16/fb_annotation/dmel_r6-16.fb_annotation'), header=T, sep = '\t')[, c('gene_symbol', 'primary_FBgn')]

# Get list of mitochondiral genes
mito <- fbgn2chrom[fbgn2chrom$chrom == 'chrM', 'FBgn']
print(length(mito))
```

### Load 10X Data
```{r load_data_function, include=FALSE}
load.data <- function(fname, name) {
    print(paste0('Running: ', name))

    tenX.data <- Read10X(data.dir = fname)
    colnames(tenX.data) <- paste(name, colnames(tenX.data), sep = "_")

    sobj <- CreateSeuratObject(
        raw.data = tenX.data,
        min.cells = 3,
        min.genes = 200,
        project = 'GroupCluster',
        display.progress = FALSE
    )
    sobj@meta.data$rep <- name
    remove(tenX.data)

    # Filter based on the number of genes
    sobj <- FilterCells(
        object = sobj,
        subset.names = "nGene",
        low.thresholds = GENE_LOW,
        high.thresholds = GENE_HIGH
    )

    # Filter based on the total UMI
    sobj <- FilterCells(
        object = sobj,
        subset.names = "nUMI",
        low.thresholds = UMI_LOW,
        high.thresholds = UMI_HIGH
    )

    sobj <- NormalizeData(
        object = sobj,
        normalization.method = "LogNormalize",
        scale.factor = 1e4,
        display.progress = FALSE
    )

    sobj <- ScaleData(
        object = sobj,
        vars.to.regress = "nUMI",
        display.progress = FALSE
    )

    sobj <- FindVariableGenes(
        object = sobj,
        do.plot = TRUE,
        display.progress = FALSE
    )
    return(sobj)
}
```

```{r load_10x_data}
sobjs <- mapply(load.data, DATA, SAMPLES, SIMPLIFY = TRUE)
```

## Find Shared Highly Variable Genes Among Replicates
```{r find_variable_genes}
find_var_genes <- function(sobj){
    var_genes <- sobj@var.genes
    return(var_genes)
}

variable_genes <- lapply(sobjs, find_var_genes)

#TODO see if this works
genes_use <- Reduce(intersect, variable_genes)
```

Variable Genes Rep 1: `r length(variable_genes[[1]])`
Variable Genes Rep 2: `r length(variable_genes[[2]])`
Variable Genes Rep 3: `r length(variable_genes[[3]])`
Variable All Three Replicates: `r length(genes_use)`

## Combine Replicates
```{r combine_replicates_w_cca}
combined <- RunMultiCCA(
    sobjs,
    genes.use = genes_use,
    num.cc = 30
)
```

```{r calculate_biocor_matrix, cache=TRUE}
biocor.mat <- MetageneBicorPlot(
    object = combined,
    grouping.var = "rep",
    dims.eval = 1:30,
    display.progress = FALSE,
    return.mat = TRUE
)
```

```{r plot_biocor, fig.width=8}
MetageneBicorPlot(
    object = combined,
    bicor.data = biocor.mat,
    grouping.var = "rep",
    dims.eval = 1:30,
    display.progress = FALSE
)
```

```{r plot_cca_dim, fig.width=8}
p1 <- DimPlot(object=combined, reduction.use='cca', group.by='rep', pt.size=0.5, do.return=TRUE)
p2 <- VlnPlot(object=combined, features.plot='CC1', group.by='rep', do.return=TRUE, point.size.use = .1)
plot_grid(p1, p2)
```

## Align Replicates
```{r align_data_w_cca, cache=TRUE}
print('Aligning Data')
combined <- AlignSubspace(
    combined,
    reduction.type = "cca",
    grouping.var = "rep",
    dims.align = 1:25
)
```

```{r plot_aligned_cca_dim, fig.width=8}
p1 <- VlnPlot(object=combined, features.plot='ACC1', group.by='rep', do.return=TRUE, point.size.use = .1)
p2 <- VlnPlot(object=combined, features.plot='ACC2', group.by='rep', do.return=TRUE, point.size.use = .1)
plot_grid(p1, p2)
```

## Cluster
```{r find_clusters, cache=TRUE}
combined <- FindClusters(
    object = combined, 
    reduction.type = "cca.aligned", 
    resolution = PARAMS, 
    dims.use = 1:25, 
    save.SNN = FALSE, 
    print.output = FALSE
)
```

```{r run_tsne, cache=TRUE}
combined <- RunTSNE(
    object = combined, 
    reduction.use = "cca.aligned", 
    dims.use = 1:25, 
    do.fast = TRUE
)
```

```{r plot_tsne_results, fig.width=8}
for (i in PARAMS){
    name <- paste0('res.', i)
    combined <- SetAllIdent(combined, id = name)
  
    p1 <- TSNEPlot(combined, do.return = T, pt.size = 0.5, group.by = "rep")
    p2 <- TSNEPlot(combined, do.label = T, do.return = T, pt.size = 0.5)
    p <- plot_grid(p1, p2)
  
    title <- ggdraw() + draw_label(name, fontface='bold')
    print(plot_grid(title, p, ncol=1, rel_heights=c(0.1, 1))) # rel_heights values control title margins
}
```

### Look at marker genes
```{r load_lit_genes}
lit_genes = yaml.load_file('../../config/literature_genes.yaml')
```
#### Spermatogonia
```{r plot_heatmap_gonia, fig.width=20, fig.height=20}
VlnPlot(object = combined, features.plot = lit_genes[['gonia']], use.raw = TRUE, y.log = TRUE)
FeaturePlot(combined, lit_genes[['gonia']], reduction.use = 'tsne', cols.use = c('gray', 'purple'))
```

#### Spermatocytes
```{r plot_heatmap_spermatocytes, fig.width=20, fig.height=20}
VlnPlot(object = combined, features.plot = lit_genes[['spermatocytes']], use.raw = TRUE, y.log = TRUE)
FeaturePlot(combined, lit_genes[['spermatocytes']], reduction.use = 'tsne', cols.use = c('gray', 'purple'))
```

#### CySC
```{r plot_heatmap_cysc, fig.width=20, fig.height=20}
VlnPlot(object = combined, features.plot = lit_genes[['cysc']], use.raw = TRUE, y.log = TRUE)
FeaturePlot(combined, lit_genes[['cysc']], reduction.use = 'tsne', cols.use = c('gray', 'purple'))
```

#### TE
```{r plot_heatmap_te, fig.width=20, fig.height=20}
VlnPlot(object = combined, features.plot = lit_genes[['te']], use.raw = TRUE, y.log = TRUE)
FeaturePlot(combined, lit_genes[['te']], reduction.use = 'tsne', cols.use = c('gray', 'purple'))
```

#### PC
```{r plot_heatmap_pc, fig.width=20}
VlnPlot(object = combined, features.plot = lit_genes[['pigment']], use.raw = TRUE, y.log = TRUE)
FeaturePlot(combined, lit_genes[['pigment']], reduction.use = 'tsne', cols.use = c('gray', 'purple'))
```

#### hub
```{r plot_heatmap_hub, fig.width=20, fig.height=20}
VlnPlot(object = combined, features.plot = lit_genes[['hub']], use.raw = TRUE, y.log = TRUE)
FeaturePlot(combined, lit_genes[['hub']], reduction.use = 'tsne', cols.use = c('gray', 'purple'))
```

### Save Results

```{r dump_seurat}
dump_seurat(object = combined, dir = OUTDIR)
```

### Differential Expression

```{r save.biomarkers}
for (i in PARAMS){
    name <- paste0('res.', i)
    combined <- SetAllIdent(combined, id = name)
  
    # Get markers ignoring replicate.
    fname <- paste0('biomarkers_', name, '.tsv')
    markers <- FindAllMarkers(object = combined, only.pos = TRUE, print.bar = FALSE)
    markers = merge(fbgn2symbol, markers, by.x='primary_FBgn', by.y='gene', all.y=T)
    save_biomarkers(markers = markers, dir = OUTDIR, fname = fname)
}
```
