```{r, include=FALSE}
knitr::opts_chunk$set(collapse=TRUE, warning=FALSE, message=FALSE,
                      bootstrap.show.code=FALSE, bootstrap.show.output=FALSE,
                      fig.ext='png')
```
# RNA-seq results

```{r imports}
library(DESeq2)
library(gridExtra)
library(ggplot2)
```

```{r child='helpers.Rmd'}
# When running interactively, run the following to load helper functions.
# When rendering this file, the contents of this code block will be skipped
# since it is a "child" block for which all contents are expected to be in
# the other file.
rmarkdown::render('helpers.Rmd', run_pandoc=FALSE)
```

```{r coldata_setup}
sample.table.filename = '../config/sampletable.tsv'

colData <- read.table(sample.table.filename, sep='\t', header=TRUE)
featurecounts.strandedness <- 's2' # minus-strand reads correspond to sense-strand transcription (e.g., TruSeq kits)

colData$featurecounts.path <- sapply(
    colData$samplename,
    function (x) file.path(
        '..', '..', 'output', 'bulk-rnaseq-wf', 'rnaseq_samples', x,
        paste0(x, '.cutadapt.bam.featurecounts', '.txt')
        )
    )

factor.columns <- c('group')

exclude.for.printing <- c('featurecounts.path', 'orig_filename')
for (col in factor.columns){
    colData[[col]] <- as.factor(colData[[col]])
}

colData$group <- relevel(colData$group, ref='OCP')
```
## Experiment overview

Here is the sample table with metadata used for this analysis:

```{r}
knitr::kable(colData[, colnames(colData)[!colnames(colData) %in% exclude.for.printing]])
```

```{r dds_initial, cache=TRUE}
# Load gene-level counts
dds <- DESeqDataSetFromFeatureCounts(
    sampleTable=colData,
    directory='/home/fearjm/Projects/larval_gonad/bulk-rnaseq-wf/downstream',
    design=~group)

dds <- dds[!grepl('ERCC-', rownames(dds)),]

# Variance-stablilized transform to normalize gene-level counts.
# Alternatively, use `rlog`, but that gives about the same results
# and is slow for large numbers of samples.
#
# Since this is for exploratory data analysis, we use blind=TRUE
# to ignore the design.
rld <- varianceStabilizingTransformation(dds, blind=TRUE)
```
## Size factors
Size factors: ideally, all libraries were sequenced to identical depth, in
which case all size factors would be 1.0. In practice, this is almost never the
case. These size factor estimates are DESeq2's way of showing how sequencing
depth varies across libraries. If some libraries are much higher or lower than
1 then those libraries had dramatically different coverage and we should be
careful about interpreting results.

```{r}
dds <- estimateSizeFactors(dds)
sf <- sizeFactors(dds)
sf <- sf[order(sf)]
knitr::kable(sf)
```

```{r dds_models, cache=TRUE}
dds <- DESeq(dds, betaPrior=FALSE)
```

## Testis w Fat vs Ovary w Fat
```{r results, cache=TRUE}
res <- results(dds, contrast = c('group', 'TF', 'OF'), alpha=.01)
res.lfc <- lfcShrink(dds, contrast = c('group', 'TF', 'OF'), res = res)
plotMA(res.lfc)
summary(res.lfc)

res.lfc$FBgn <- rownames(res)
res.lfc <- res.lfc[,c('FBgn', 'baseMean', 'log2FoldChange', 'lfcSE', 'stat', 'pvalue', 'padj')]
write.table(res.lfc, '../../output/bulk-rnaseq-wf/deseq2_results_tf_vs_of.tsv', sep='\t', row.names = FALSE, quote=FALSE)
```


## Testis cleaned w papain vs Ovary cleaned w papain
```{r results2, cache=TRUE}
res <- results(dds, contrast = c('group', 'TCP', 'OCP'), alpha=.01)
res.lfc <- lfcShrink(dds, contrast = c('group', 'TCP', 'OCP'), res = res)
plotMA(res.lfc)
summary(res.lfc)

res.lfc$FBgn <- rownames(res)
res.lfc <- res.lfc[,c('FBgn', 'baseMean', 'log2FoldChange', 'lfcSE', 'stat', 'pvalue', 'padj')]
write.table(res.lfc, '../../output/bulk-rnaseq-wf/deseq2_results_tcp_vs_ocp.tsv', sep='\t', row.names = FALSE, quote=FALSE)
```

## Testis dissociated w papain vs Ovary dissociated w papain
*NOTE* F9_TDP had a really large size factor. 
```{r results3, cache=TRUE}
res <- results(dds, contrast = c('group', 'TDP', 'ODP'), alpha=.01)
res.lfc <- lfcShrink(dds, contrast = c('group', 'TDP', 'ODP'), res = res)
plotMA(res.lfc)
summary(res.lfc)

res.lfc$FBgn <- rownames(res)
res.lfc <- res.lfc[,c('FBgn', 'baseMean', 'log2FoldChange', 'lfcSE', 'stat', 'pvalue', 'padj')]
write.table(res.lfc, '../../output/bulk-rnaseq-wf/deseq2_results_tdp_vs_odp.tsv', sep='\t', row.names = FALSE, quote=FALSE)
```

# Session info
For reproducibility purposes, here is the output of `sessionInfo()` showing the
versions of all packages used here.

```{r, collapse=FALSE}
sessionInfo()
```

# Help

```{r helpdocs, child="help_docs.Rmd", run_pandoc=FALSE}
# NOTE: optional help section
#   Delete this chunk, or set to eval=FALSE, if you don't want to include the
#   help text from "help_docs.Rmd"
```
