---
title: Germ Cells Differential Expression
author: Justin Fear
---

```{r, include=FALSE}
knitr::opts_chunk$set(collapse=TRUE, warning=FALSE, message=FALSE, cache.lazy=FALSE,
                      bootstrap.show.code=FALSE, bootstrap.show.output=FALSE,
                      fig.ext='png')
```


Here I focus on the 5 germ cell clusters {Spermatogonia, Early, Mid, and Late
Spermatocytes, Cluster 11}. The goal is to figure out what are the differences
between these groups of cells.

I am assuming the following annotations:

* 0: 'Late 1ยบ Spermatocytes (0)'
* 1: 'Mid Cyst Cells (1)'
* 2: 'Mid 1ยบ Spermatocytes (2)'
* 3: 'Early 1ยบ Spermatocytes (3)'
* 4: 'Late Cyst Cells (4)'
* 5: 'Early Cyst Cells (5)'
* 6: 'Spermatogonia (6)'
* 7: 'Terminal Epithelium (7)'
* 8: 'Pigment Cells (8)'
* 9: 'Unknown (9)'
* 10: 'Unknown (10)'
* 11: 'Unknown (11)'

```{r load_libs}
library(Seurat)
library(dplyr)
source('../../lib/seurat.R')

REFERENCES_DIR <- Sys.getenv('REFERENCES_DIR', '/data/LCDB/lcdb-references')
OUTDIR <- '../../output/scrnaseq-wf/germcell_deg'
fbgn2symbol <- read.csv(
  file.path(REFERENCES_DIR,
    'dmel/r6-16/fb_annotation/dmel_r6-16.fb_annotation'
  ),
  header=T,
  sep = '\t'
)[, c('gene_symbol', 'primary_FBgn')]
```

```{r load.data}
load('../../output/scrnaseq-wf/scrnaseq_combine_force/seurat.Robj')
sobj <- object
sobj <- SetAllIdent(sobj, id='res.0.6')
```

```{r gonia.cytes}
markers <- FindMarkers(sobj, ident.1 = 6, ident.2 = c(3, 2, 0))
fname <- 'gonia_vs_cytes.tsv'
markers$gene = rownames(markers)
markers = merge(fbgn2symbol, markers, by.x='primary_FBgn', by.y='gene', all.y=T)
save_biomarkers(markers = markers, dir = OUTDIR, fname = fname)
```

```{r gonia.cytes11}
markers <- FindMarkers(sobj, ident.1 = 6, ident.2 = c(3, 2, 0, 11))
fname <- 'gonia_vs_cytes_plus_11.tsv'
markers$gene = rownames(markers)
markers = merge(fbgn2symbol, markers, by.x='primary_FBgn', by.y='gene', all.y=T)
save_biomarkers(markers = markers, dir = OUTDIR, fname = fname)
```

```{r gonia.early}
markers <- FindMarkers(sobj, ident.1 = 6, ident.2 = 3)
fname <- 'gonia_vs_early.tsv'
markers$gene = rownames(markers)
markers = merge(fbgn2symbol, markers, by.x='primary_FBgn', by.y='gene', all.y=T)
save_biomarkers(markers = markers, dir = OUTDIR, fname = fname)
```

```{r gonia.mid}
markers <- FindMarkers(sobj, ident.1 = 6, ident.2 = 2)
fname <- 'gonia_vs_mid.tsv'
markers$gene = rownames(markers)
markers = merge(fbgn2symbol, markers, by.x='primary_FBgn', by.y='gene', all.y=T)
save_biomarkers(markers = markers, dir = OUTDIR, fname = fname)
```

```{r gonia.late}
markers <- FindMarkers(sobj, ident.1 = 6, ident.2 = 0)
fname <- 'gonia_vs_late.tsv'
markers$gene = rownames(markers)
markers = merge(fbgn2symbol, markers, by.x='primary_FBgn', by.y='gene', all.y=T)
save_biomarkers(markers = markers, dir = OUTDIR, fname = fname)
```

```{r gonia.eleven}
markers <- FindMarkers(sobj, ident.1 = 6, ident.2 = 11)
fname <- 'gonia_vs_eleven.tsv'
markers$gene = rownames(markers)
markers = merge(fbgn2symbol, markers, by.x='primary_FBgn', by.y='gene', all.y=T)
save_biomarkers(markers = markers, dir = OUTDIR, fname = fname)
```

```{r early.mid}
markers <- FindMarkers(sobj, ident.1 = 3, ident.2 = 2)
fname <- 'early_vs_mid.tsv'
markers$gene = rownames(markers)
markers = merge(fbgn2symbol, markers, by.x='primary_FBgn', by.y='gene', all.y=T)
save_biomarkers(markers = markers, dir = OUTDIR, fname = fname)
```

```{r early.late}
markers <- FindMarkers(sobj, ident.1 = 3, ident.2 = 0)
fname <- 'early_vs_late.tsv'
markers$gene = rownames(markers)
markers = merge(fbgn2symbol, markers, by.x='primary_FBgn', by.y='gene', all.y=T)
save_biomarkers(markers = markers, dir = OUTDIR, fname = fname)
```

```{r early.eleven}
markers <- FindMarkers(sobj, ident.1 = 3, ident.2 = 11)
fname <- 'early_vs_eleven.tsv'
markers$gene = rownames(markers)
markers = merge(fbgn2symbol, markers, by.x='primary_FBgn', by.y='gene', all.y=T)
save_biomarkers(markers = markers, dir = OUTDIR, fname = fname)
```

```{r mid.late}
markers <- FindMarkers(sobj, ident.1 = 2, ident.2 = 0)
fname <- 'mid_vs_late.tsv'
markers$gene = rownames(markers)
markers = merge(fbgn2symbol, markers, by.x='primary_FBgn', by.y='gene', all.y=T)
save_biomarkers(markers = markers, dir = OUTDIR, fname = fname)
```

```{r mid.eleven}
markers <- FindMarkers(sobj, ident.1 = 2, ident.2 = 11)
fname <- 'mid_vs_eleven.tsv'
markers$gene = rownames(markers)
markers = merge(fbgn2symbol, markers, by.x='primary_FBgn', by.y='gene', all.y=T)
save_biomarkers(markers = markers, dir = OUTDIR, fname = fname)
```

```{r late.eleven}
markers <- FindMarkers(sobj, ident.1 = 0, ident.2 = 11)
fname <- 'late_vs_eleven.tsv'
markers$gene = rownames(markers)
markers = merge(fbgn2symbol, markers, by.x='primary_FBgn', by.y='gene', all.y=T)
save_biomarkers(markers = markers, dir = OUTDIR, fname = fname)
```

