{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Identifying Cell Types of the Testis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import re\n",
    "from pathlib import Path\n",
    "from yaml import load\n",
    "\n",
    "from IPython.display import display, HTML, Markdown\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load all of my config settings\n",
    "config = {}\n",
    "with open('../config/common.yaml') as fh:\n",
    "    config.update(load(fh.read()))\n",
    "    \n",
    "with open('../config/colors.yaml') as fh:\n",
    "    config['colors'] = load(fh.read())\n",
    "    \n",
    "with open('../science_submission/config.yaml') as fh:\n",
    "    config.update(load(fh.read()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "72\n",
      "abd-A, Abd-B, aly, apt, bam, bgcn, bnb, bol, bw, c-cup, can, CG11697, CG18628, CG3927, CG8368, d-cup, dj, EcR, Eip93F, ems, eya, fax, foxo, fzo, gbb, glob1, hui, ImpL2, kek1, Marf, mia, mle, Mst87F, MtnA, N, neur, nht, nord, nos, Nrt, ocn, oys, p-cup, p53, peb, Phf7, puc, r-cup, Rbp9, retn, robo2, sa, sano, sev, Six4, so, soti, sowi, Sox100B, spict, sunz, tej, tj, topi, tut, vas, vkg, vn, wa-cup, Wnt2, Wnt4, zfh1\n"
     ]
    }
   ],
   "source": [
    "# Get list of genes from the literature\n",
    "symbol2fbgn = pd.read_pickle('../output/science_submission/symbol2fbgn.pkl')\n",
    "lit_genes = config['lit_genes_long']\n",
    "lit_genes_fbgn = [symbol2fbgn[x] for x in config['lit_genes_long']]\n",
    "print(len(lit_genes))\n",
    "print(', '.join(sorted(lit_genes, key=lambda x: x.lower())))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Cells Per Cluster</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SP</th>\n",
       "      <td>1,367</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E1º</th>\n",
       "      <td>1,418</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M1º</th>\n",
       "      <td>1,651</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L1º</th>\n",
       "      <td>2,014</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC</th>\n",
       "      <td>1,373</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MC</th>\n",
       "      <td>1,744</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LC</th>\n",
       "      <td>1,384</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TE</th>\n",
       "      <td>946</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PC</th>\n",
       "      <td>707</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Cells Per Cluster\n",
       "SP              1,367\n",
       "E1º             1,418\n",
       "M1º             1,651\n",
       "L1º             2,014\n",
       "EC              1,373\n",
       "MC              1,744\n",
       "LC              1,384\n",
       "TE                946\n",
       "PC                707"
      ]
     },
     "execution_count": 152,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get mappint of cell_id to short cluster name, remove unknown clusters\n",
    "clusters = (\n",
    "    pd.read_parquet('../output/scrnaseq-wf/clusters.parquet')\n",
    "    .assign(cluster = lambda df: pd.Categorical(df.cluster.map(config['short_cluster_annot']), ordered=True, categories=config['short_cluster_order']))\n",
    "    .dropna()\n",
    ")\n",
    "clusters.cluster.value_counts().sort_index().map(lambda x: f'{x:,}').rename('Cells Per Cluster').to_frame()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Biomarkers per cluster</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cluster</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SP</th>\n",
       "      <td>1,383</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E1º</th>\n",
       "      <td>1,417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M1º</th>\n",
       "      <td>1,744</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L1º</th>\n",
       "      <td>760</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC</th>\n",
       "      <td>616</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MC</th>\n",
       "      <td>352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LC</th>\n",
       "      <td>583</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TE</th>\n",
       "      <td>638</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PC</th>\n",
       "      <td>733</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Biomarkers per cluster\n",
       "cluster                       \n",
       "SP                       1,383\n",
       "E1º                      1,417\n",
       "M1º                      1,744\n",
       "L1º                        760\n",
       "EC                         616\n",
       "MC                         352\n",
       "LC                         583\n",
       "TE                         638\n",
       "PC                         733"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 4,872 biomakrer genes\n"
     ]
    }
   ],
   "source": [
    "# Get biomarkers\n",
    "resolution = config['resolution']\n",
    "biomarkers = (\n",
    "    pd.read_csv(f'../output/scrnaseq-wf/scrnaseq_combine_force/biomarkers_{resolution}.tsv', sep='\\t', index_col=0)\n",
    "    .rename_axis('FBgn')\n",
    "    .query('p_val_adj <= 0.05')\n",
    "    .assign(cluster = lambda df: pd.Categorical(df.cluster.map(config['short_cluster_annot']), ordered=True, categories=config['short_cluster_order']))\n",
    "    .dropna()\n",
    "    .loc[:, ['gene_symbol', 'cluster']]\n",
    ")\n",
    "display(biomarkers.groupby('cluster').size().sort_index().map(lambda x: f'{x:,}').rename('Biomarkers per cluster').to_frame())\n",
    "print(f'There are {biomarkers.index.unique().shape[0]:,} biomakrer genes')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Of the 72 literature genes, only 56 where in the biomarker list.\n"
     ]
    }
   ],
   "source": [
    "print(f'Of the {len(lit_genes_fbgn):,} literature genes, only {biomarkers.query(f\"FBgn == {lit_genes_fbgn}\").index.unique().shape[0]:,} where in the biomarker list.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<h3>Spermatogonia</h3>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bam, bgcn, CG11697, Marf, nos, p53, peb, Phf7, Rbp9, tej, tut, vas\n",
      "There were (5 / 7 = 71.43%) literature genes that were called biomarker in the correct cluster.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>clusters</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_symbol</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>bam</th>\n",
       "      <td>SP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CG11697</th>\n",
       "      <td>E1º|M1º|L1º</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p53</th>\n",
       "      <td>SP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>peb</th>\n",
       "      <td>LC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Phf7</th>\n",
       "      <td>SP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rbp9</th>\n",
       "      <td>SP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>vas</th>\n",
       "      <td>SP</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                clusters\n",
       "gene_symbol             \n",
       "bam                   SP\n",
       "CG11697      E1º|M1º|L1º\n",
       "p53                   SP\n",
       "peb                   LC\n",
       "Phf7                  SP\n",
       "Rbp9                  SP\n",
       "vas                   SP"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<h3>Primary Spermatocytes</h3>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "aly, bol, c-cup, can, CG3927, d-cup, dj, fzo, mia, mle, Mst87F, nht, ocn, oys, p-cup, r-cup, sa, soti, sowi, sunz, topi, wa-cup\n",
      "There were (18 / 18 = 100.00%) literature genes that were called biomarker in the correct cluster.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>clusters</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_symbol</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>aly</th>\n",
       "      <td>E1º</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bol</th>\n",
       "      <td>E1º|M1º</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c-cup</th>\n",
       "      <td>E1º|M1º|L1º</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>can</th>\n",
       "      <td>E1º</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CG3927</th>\n",
       "      <td>E1º|M1º|L1º</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d-cup</th>\n",
       "      <td>M1º</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dj</th>\n",
       "      <td>E1º|M1º|L1º</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>fzo</th>\n",
       "      <td>M1º</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mst87F</th>\n",
       "      <td>E1º|M1º|L1º</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nht</th>\n",
       "      <td>SP|E1º</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ocn</th>\n",
       "      <td>E1º|M1º|L1º</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>oys</th>\n",
       "      <td>E1º</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sa</th>\n",
       "      <td>E1º</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>soti</th>\n",
       "      <td>E1º|M1º|L1º</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sowi</th>\n",
       "      <td>E1º|M1º|L1º</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sunz</th>\n",
       "      <td>M1º</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>topi</th>\n",
       "      <td>E1º</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>wa-cup</th>\n",
       "      <td>E1º|M1º</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                clusters\n",
       "gene_symbol             \n",
       "aly                  E1º\n",
       "bol              E1º|M1º\n",
       "c-cup        E1º|M1º|L1º\n",
       "can                  E1º\n",
       "CG3927       E1º|M1º|L1º\n",
       "d-cup                M1º\n",
       "dj           E1º|M1º|L1º\n",
       "fzo                  M1º\n",
       "Mst87F       E1º|M1º|L1º\n",
       "nht               SP|E1º\n",
       "ocn          E1º|M1º|L1º\n",
       "oys                  E1º\n",
       "sa                   E1º\n",
       "soti         E1º|M1º|L1º\n",
       "sowi         E1º|M1º|L1º\n",
       "sunz                 M1º\n",
       "topi                 E1º\n",
       "wa-cup           E1º|M1º"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<h3>Somatic Cyst Cells</h3>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "apt, bnb, CG8368, EcR, Eip93F, eya, fax, foxo, gbb, glob1, hui, ImpL2, kek1, neur, Nrt, puc, robo2, sano, sev, spict, tj, vn, Wnt4, zfh1\n",
      "There were (12 / 19 = 63.16%) literature genes that were called biomarker in the correct cluster.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>clusters</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_symbol</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>bnb</th>\n",
       "      <td>EC|LC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CG8368</th>\n",
       "      <td>E1º|M1º</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EcR</th>\n",
       "      <td>TE|PC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Eip93F</th>\n",
       "      <td>LC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>fax</th>\n",
       "      <td>EC|MC|LC|TE|PC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>foxo</th>\n",
       "      <td>EC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gbb</th>\n",
       "      <td>TE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>glob1</th>\n",
       "      <td>MC|LC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hui</th>\n",
       "      <td>EC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ImpL2</th>\n",
       "      <td>EC|TE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>kek1</th>\n",
       "      <td>EC|TE|PC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Nrt</th>\n",
       "      <td>EC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>puc</th>\n",
       "      <td>TE|PC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>robo2</th>\n",
       "      <td>TE|PC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>spict</th>\n",
       "      <td>PC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tj</th>\n",
       "      <td>EC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>vn</th>\n",
       "      <td>EC|PC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Wnt4</th>\n",
       "      <td>EC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>zfh1</th>\n",
       "      <td>TE|PC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   clusters\n",
       "gene_symbol                \n",
       "bnb                   EC|LC\n",
       "CG8368              E1º|M1º\n",
       "EcR                   TE|PC\n",
       "Eip93F                   LC\n",
       "fax          EC|MC|LC|TE|PC\n",
       "foxo                     EC\n",
       "gbb                      TE\n",
       "glob1                 MC|LC\n",
       "hui                      EC\n",
       "ImpL2                 EC|TE\n",
       "kek1               EC|TE|PC\n",
       "Nrt                      EC\n",
       "puc                   TE|PC\n",
       "robo2                 TE|PC\n",
       "spict                    PC\n",
       "tj                       EC\n",
       "vn                    EC|PC\n",
       "Wnt4                     EC\n",
       "zfh1                  TE|PC"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<h3>Terminal Epithelium</h3>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "abd-A, Abd-B, CG18628, MtnA, nord, retn, Six4, so, Wnt2\n",
      "There were (5 / 7 = 71.43%) literature genes that were called biomarker in the correct cluster.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>clusters</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_symbol</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>abd-A</th>\n",
       "      <td>TE|PC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Abd-B</th>\n",
       "      <td>TE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MtnA</th>\n",
       "      <td>EC|MC|LC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nord</th>\n",
       "      <td>TE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>retn</th>\n",
       "      <td>TE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Six4</th>\n",
       "      <td>EC|MC|LC|TE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Wnt2</th>\n",
       "      <td>MC|LC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                clusters\n",
       "gene_symbol             \n",
       "abd-A              TE|PC\n",
       "Abd-B                 TE\n",
       "MtnA            EC|MC|LC\n",
       "nord                  TE\n",
       "retn                  TE\n",
       "Six4         EC|MC|LC|TE\n",
       "Wnt2               MC|LC"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<h3>Pigment Cells</h3>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bw, N, Sox100B, vkg\n",
      "There were (4 / 4 = 100.00%) literature genes that were called biomarker in the correct cluster.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>clusters</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_symbol</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>bw</th>\n",
       "      <td>PC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N</th>\n",
       "      <td>EC|TE|PC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sox100B</th>\n",
       "      <td>PC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>vkg</th>\n",
       "      <td>PC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             clusters\n",
       "gene_symbol          \n",
       "bw                 PC\n",
       "N            EC|TE|PC\n",
       "Sox100B            PC\n",
       "vkg                PC"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "44 / 55 = 0.80%\n"
     ]
    }
   ],
   "source": [
    "# Figure out how many literature genes were called biomarker for the right cluster.\n",
    "PASS = 0\n",
    "TOTAL = 0\n",
    "def get_lit_gene_subset(cell_type):\n",
    "    if cell_type == 'gonia':\n",
    "        name = 'Spermatogonia'\n",
    "        idx = (0, 12)\n",
    "    elif cell_type == 'primary':\n",
    "        name = 'Primary Spermatocytes'\n",
    "        idx = (12, 34)\n",
    "    elif cell_type == 'cyst':\n",
    "        name = 'Somatic Cyst Cells'\n",
    "        idx = (34, 58)\n",
    "    elif cell_type == 'te':\n",
    "        name = 'Terminal Epithelium'\n",
    "        idx = (58, 67)\n",
    "    elif cell_type == 'pc':\n",
    "        name = 'Pigment Cells'\n",
    "        idx = (67,71)\n",
    "    display(HTML(f'<h3>{name}</h3>'))\n",
    "    print(', '.join(sorted(lit_genes[idx[0]:idx[1]], key=lambda x: x.lower())))\n",
    "    return lit_genes_fbgn[idx[0]: idx[1]]\n",
    "\n",
    "def check_biomarkers(fbgns, cell_type_pattern):\n",
    "    subset = (\n",
    "        biomarkers.query(f'FBgn == {fbgns}')\n",
    "        .groupby('gene_symbol')\n",
    "        .apply(lambda df: '|'.join(df.cluster.sort_values().values))\n",
    "        .rename('clusters')\n",
    "        .to_frame()\n",
    "        .assign(lower=lambda df: df.index.str.lower())\n",
    "        .sort_values(by='lower')\n",
    "        .drop('lower', axis=1)\n",
    "    )\n",
    "    num_with_correct_cell_type = subset.clusters.str.contains(cell_type_pattern).sum()\n",
    "    total_num_genes = subset.shape[0]\n",
    "    \n",
    "    global PASS\n",
    "    global TOTAL\n",
    "    PASS += num_with_correct_cell_type\n",
    "    TOTAL += total_num_genes\n",
    "    \n",
    "    print(f'There were ({num_with_correct_cell_type:,} / {total_num_genes:,} = {num_with_correct_cell_type / total_num_genes * 100:.2f}%) literature genes that were called biomarker in the correct cluster.')\n",
    "    display(subset)\n",
    "\n",
    "fbgns = get_lit_gene_subset('gonia')\n",
    "check_biomarkers(fbgns, 'SP')\n",
    "\n",
    "fbgns = get_lit_gene_subset('primary')\n",
    "check_biomarkers(fbgns, 'E1º|M1º|L1º')\n",
    "\n",
    "fbgns = get_lit_gene_subset('cyst')\n",
    "check_biomarkers(fbgns, 'EC|MC|LC')\n",
    "\n",
    "fbgns = get_lit_gene_subset('te')\n",
    "check_biomarkers(fbgns, 'TE')\n",
    "\n",
    "fbgns = get_lit_gene_subset('pc')\n",
    "check_biomarkers(fbgns, 'PC')\n",
    "\n",
    "print(f'{PASS:,} / {TOTAL:,} = {PASS / TOTAL:,.2f}%')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "jupytext": {
   "main_language": "python",
   "text_representation": {
    "extension": ".py",
    "format_name": "percent",
    "format_version": "1.2",
    "jupytext_version": "1.0.0"
   }
  },
  "kernelspec": {
   "display_name": "Python [conda env:larval_gonad]",
   "language": "python",
   "name": "conda-env-larval_gonad-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
