---
title: Initial Monocle Pseudotime Analysis
author: Justin Fear
date: 'Compiled: `r format(Sys.Date(), "%B %d, %Y")`'
---

```{r, include=FALSE}
knitr::opts_chunk$set(collapse=TRUE, warning=FALSE, message=FALSE, cache.lazy=FALSE,
                      bootstrap.show.code=FALSE, bootstrap.show.output=FALSE,
                      fig.ext='png')
```

```{r setup}
options(repr.plot.width=10, repr.plot.height=10)
library('monocle')
library('dplyr')
```

```{r, get.gene.annot}
# create a dataframe with FBgn to gene symbol
genes <- read.table(file.path(Sys.getenv('REFERENCES_DIR'), 'dmel/r6-16/fb_annotation/dmel_r6-16.fb_annotation'),
		    sep='\t',
		    stringsAsFactors=F,
		    header=T,
		    quote=NULL,
		    colClasses=c(rep("character", 2), rep("NULL", 3))
		    )
row.names(genes) <- genes$primary_FBgn
genes <- subset(genes, select=-primary_FBgn)
```


```{r load.seurat}
# Load data from seurat object.
load('../output/scrnaseq-wf/scrnaseq_combine_force/seurat.Robj')
sobj <- object
dds <- importCDS(sobj)
dds <- updateCDS(dds)
```
I am assuming the following annotations:

* 0: 'Late 1ยบ Spermatocytes (0)'
* 1: 'Mid Cyst Cells (1)'
* 2: 'Mid 1ยบ Spermatocytes (2)'
* 3: 'Early 1ยบ Spermatocytes (3)'
* 4: 'Late Cyst Cells (4)'
* 5: 'Early Cyst Cells (5)'
* 6: 'Spermatogonia (6)'
* 7: 'Terminal Epithelium (7)'
* 8: 'Pigment Cells (8)'
* 9: 'Unknown (9)'
* 10: 'Unknown (10)'
* 11: 'Unknown (11)'

```{r}
lvls = c(
  '0' = 'L1',
  '1' = 'MC',
  '2' = 'M1',
  '3' = 'E1',
  '4' = 'LC',
  '5' = 'EC',
  '6' = 'SP',
  '7' = 'TE',
  '8' = 'PC',
  '9' = 'U1',
  '10' = 'U2',
  '11' = 'U3'
)

pData(dds)$cell_type <- plyr::revalue(as.character(pData(dds)$res.0.6), lvls)

cell_type_color <- c(
  "SP" = "#b11218",
  "E1" = "#e32f27",
  "M1" = "#fb6b4b",
  "L1" = "#fdd4c2",
  "EC" = "#105ba4",
  "MC" = "#3787c0",
  "LC" = "#6caed6",
  "TE" = "#6a3d9a",
  "PC" = "#b15928",
  "U1" = "#d1d1d1",
  "U2" = "#d1d1d1",
  "U3" = "#000000"
)

```


```{r}
DelayedArray:::set_verbose_block_processing(TRUE)
options(DelayedArray.auto.block.size = 1000e6)

dds <- estimateSizeFactors(dds)
dds <- estimateDispersions(dds)
```

```{r}
dds <- preprocessCDS(dds, num_dim = 20)
dds <- reduceDimension(dds, reduction_method = 'UMAP')
```

```{r}
dds <- partitionCells(dds)
```

```{r}
dds <- learnGraph(dds, RGE_method = 'SimplePPT')
```

```{r fig.width=10}
plot_cell_trajectory(dds, color_by = "cell_type") + scale_color_manual(values = cell_type_color)
```


```{r}

```





```{r get.germ}
# Pull out only the germ cells based on res.0.6
germ <- row.names(subset(pData(dds), res.0.6 %in% c(0, 1, 2, 3, 4, 5, 6, 7, 8)))
dds_germ <- dds[, germ]
```

```{r add.gene.annot}
# Add on gene symbols to make easier to read
merged <- merge(fData(dds_germ), genes, by='row.names', all.x = T, all.y = F)
row.names(merged) <- merged$Row.names
merged <- subset(merged, select=-c(Row.names, gene_short_name))
colnames(merged) <- c('gene_short_name')

fData(dds_germ) <- merged[row.names(fData(dds_germ)), , drop=F]
```

```{r estimate.norms}
# Detect expressed genes and normalize
dds_germ <- detectGenes(dds_germ, min_expr = 0.1)
dds_germ <- estimateSizeFactors(dds_germ)
dds_germ <- estimateDispersions(dds_germ)
```

```{r deg.select.for.order}
# Create a list of genes for ordering trajectory.
# (1) Filter genes in a few cells
# (2) run the differential expression analysis
# (3) selecte significant genes for use as ordering
# (4) filter genes not in this list.

# TODO: currently I am using clusters called by seurat, but it is probably worth
# calling clusters again on just this subset of data.
expressed_genes <- row.names(subset(fData(dds_germ), num_cells_expressed >= 10))
diff_test_res <- differentialGeneTest(dds_germ[expressed_genes,], fullModelFormulaStr = "~res.0.6")
ordering_genes = row.names(subset(diff_test_res, qval < 0.01))
dds_order = setOrderingFilter(dds_germ, ordering_genes)
```

```{r}
plot_ordering_genes(dds_order)
```

```{r order.cells}
dds_order <- reduceDimension(dds_order, max_components = 2, method = 'DDRTree')
dds_order <- orderCells(dds_order)
```

```{r plot.trajectory, fig.width=20}
#png('../output/pseudotime-wf/rough_trajectory.png', width = 1600, height = 1600, res = 150)
plot_cell_trajectory(dds_order, color_by = 'res.0.6') + scale_color_brewer(palette="Paired")
#dev.off()
```

```{r output}
save(dds_order, file = '../output/pseudotime-wf/monocle.Robj')
load('../output/pseudotime-wf/monocle.Robj')
```

```{r dump.obj}
# Save useful data for python plotting

# Cell location
cell.loc <- t(reducedDimS(dds_order))
colnames(cell.loc) <- c('C1', 'C2')
cell.loc <- as.data.frame(cell.loc) %>% tibble::rownames_to_column('cell_id')
write.table(cell.loc, file = '../output/pseudotime-wf/cell_location.tsv', sep = '\t', quote = F, row.names = F)

# Tree location
tree.loc <- t(reducedDimK(dds_order))
colnames(tree.loc) <- c('C1', 'C2')
tree.loc <- as.data.frame(tree.loc) %>% tibble::rownames_to_column('tree_id')
write.table(tree.loc, file = '../output/pseudotime-wf/tree_location.tsv', sep = '\t', quote = F, row.names = F)

branch.point <- dds_order@auxOrderingData[['DDRTree']]$branch_points
write(branch.point, file = '../output/pseudotime-wf/branch_points.txt', sep = '\n')

metadata <- pData(dds_order)[c('Size_Factor', 'num_genes_expressed', 'Pseudotime', 'State')]
metadata <- as.data.frame(metadata) %>% tibble::rownames_to_column('cell_id')
write.table(metadata, file = '../output/pseudotime-wf/pseudotime.tsv', sep = '\t', quote = F, row.names = F)

```

